server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java

```diff
@@ -0,0 +1,176 @@
+import org.sonar.core.persistence.MyBatis;
+import org.sonar.server.db.DbClient;
+import org.sonar.server.exceptions.ServerException;
+import org.sonar.server.user.UserSession;
+
+public class CreateAction implements MetricsWsAction {
+  private static final String ACTION = "create";
+  public static final String PARAM_NAME = "name";
+  public static final String PARAM_KEY = "key";
+  public static final String PARAM_TYPE = "type";
+  public static final String PARAM_DESCRIPTION = "description";
+  public static final String PARAM_DOMAIN = "domain";
+
+  private static final String FIELD_ID = "id";
+  private static final String FIELD_NAME = PARAM_NAME;
+  private static final String FIELD_KEY = PARAM_KEY;
+  private static final String FIELD_TYPE = PARAM_TYPE;
+  private static final String FIELD_DESCRIPTION = PARAM_DESCRIPTION;
+  private static final String FIELD_DOMAIN = PARAM_DOMAIN;
+
+  private final DbClient dbClient;
+  private final UserSession userSession;
+
+  public CreateAction(DbClient dbClient, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION)
+      .setPost(true)
+      .setDescription("Create custom metric.<br /> Requires 'Administer System' permission.")
+      .setSince("5.2")
+      .setHandler(this);
+
+    action.createParam(PARAM_NAME)
+      .setRequired(true)
+      .setDescription("Name")
+      .setExampleValue("Team Size");
+
+    action.createParam(PARAM_KEY)
+      .setRequired(true)
+      .setDescription("Key")
+      .setExampleValue("team_size");
+
+    action.createParam(PARAM_TYPE)
+      .setRequired(true)
+      .setDescription("Metric type key")
+      .setPossibleValues(Metric.ValueType.names())
+      .setExampleValue(String.format("<ul><li>%s</li><li>%s</li></ul>", Metric.ValueType.INT.name(), Metric.ValueType.BOOL.name()));
+
+    action.createParam(PARAM_DESCRIPTION)
+      .setDescription("Description")
+      .setExampleValue("Size of the team");
+
+    action.createParam(PARAM_DOMAIN)
+      .setDescription("Domain")
+      .setExampleValue("Tests");
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkLoggedIn().checkGlobalPermission(GlobalPermissions.SYSTEM_ADMIN);
+    String key = request.mandatoryParam(PARAM_KEY);
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      MetricDto metricInDb = dbClient.metricDao().selectNullableByKey(dbSession, key);
+      checkMetricInDb(metricInDb, key);
+
+      MetricDto metricTemplate = new MetricDto()
+        .setKey(key)
+        .setShortName(request.mandatoryParam(PARAM_NAME))
+        .setValueType(Metric.ValueType.valueOf(request.mandatoryParam(PARAM_TYPE)).name())
+        .setDescription(request.param(PARAM_DESCRIPTION))
+        .setDomain(request.param(PARAM_DOMAIN));
```

Comment:
nope, clearer to read but I changed it to set when it's not null. I still find this rule cumbersome for dto. By default field are null so it makes sense to set directly. In the end the code is less readable but more explicit.